File: backend/app/schemas/candidate.py
"""Candidate Pydantic schemas"""

from typing import List, Optional, Any
from uuid import UUID
from datetime import datetime, date
from pydantic import BaseModel, EmailStr, Field, model_validator

# Import related schemas (forward compatibility)
from app.schemas.candidate_screening import CandidateScreeningResponse
from app.schemas.candidate_document import CandidateDocumentResponse
from app.schemas.candidate_counseling import CandidateCounselingResponse


# Nested Schemas for JSON fields


class Degree(BaseModel):
    degree_name: str
    specialization: str
    college_name: str
    year_of_passing: int
    percentage: float


class EducationDetails(BaseModel):
    degrees: List[Degree] = []


class WorkExperience(BaseModel):
    is_experienced: bool = False
    currently_employed: bool = False
    year_of_experience: Optional[str] = None


class DisabilityDetails(BaseModel):
    is_disabled: bool = False
    disability_type: Optional[str] = None
    disability_percentage: Optional[float] = None


# CRUD Schemas

class CandidateBase(BaseModel):
    name: str
    gender: str
    email: EmailStr
    phone: str
    whatsapp_number: Optional[str] = None
    dob: Optional[date] = None
    pincode: str
    city: Optional[str] = None
    district: Optional[str] = None
    state: Optional[str] = None
    country_code: str = Field(default="IN", description="Country code for address validation")
    
    guardian_details: Optional[dict] = None
    work_experience: Optional[WorkExperience] = None
    education_details: Optional[EducationDetails] = None
    disability_details: Optional[DisabilityDetails] = None


class CandidateCreate(CandidateBase):
    pass


class CandidateCheck(BaseModel):
    email: EmailStr
    phone: str
    pincode: str
    city: Optional[str] = None
    district: Optional[str] = None
    state: Optional[str] = None
    country_code: str = "IN"


class CandidateUpdate(BaseModel):
    name: Optional[str] = None
    gender: Optional[str] = None
    email: Optional[EmailStr] = None
    phone: Optional[str] = None
    whatsapp_number: Optional[str] = None
    dob: Optional[date] = None
    pincode: Optional[str] = None
    city: Optional[str] = None
    district: Optional[str] = None
    state: Optional[str] = None
    country_code: Optional[str] = "IN"
    guardian_details: Optional[dict] = None
    work_experience: Optional[dict] = None
    education_details: Optional[EducationDetails] = None
    disability_details: Optional[DisabilityDetails] = None


class CandidateResponse(CandidateBase):
    """
    Candidate response schema.
    Note: Uses public_id (UUID) instead of internal id for security.
    """
    id: int
    public_id: UUID  # Secure UUID for external API
    city: str
    district: str
    state: str
    created_at: datetime
    updated_at: datetime
    
    # Optional nested relationships (filled by trainers)
    screening: Optional[CandidateScreeningResponse] = None
    documents: List[CandidateDocumentResponse] = []
    counseling: Optional[CandidateCounselingResponse] = None
    
    @classmethod
    def model_validate(cls, obj, *args, **kwargs):
        """Custom validation to handle unloaded SQLAlchemy relationships"""
        from sqlalchemy.orm import object_session
        from sqlalchemy.inspect import inspect
        
        # If this is a SQLAlchemy object, check which relationships are loaded
        if hasattr(obj, '__dict__'):
            state = inspect(obj)
            if state:
                # Only include relationships that are actually loaded
                data = {}
                for key in cls.model_fields.keys():
                    if key in ['screening', 'documents', 'counseling']:
                        # Check if the relationship is loaded
                        if key in state.unloaded:
                            # Skip unloaded relationships
                            continue
                    # Get the attribute value
                    try:
                        data[key] = getattr(obj, key)
                    except:
                        # If we can't get it, skip it
                        continue
                        
                return super().model_validate(data, *args, **kwargs)
        
        return super().model_validate(obj, *args, **kwargs)
    
    class Config:
        from_attributes = True


class CandidateListResponse(BaseModel):
    """Simplified response for list endpoints"""
    id: int
    public_id: UUID
    name: str
    gender: str
    email: Any
    phone: str
    whatsapp_number: Optional[str] = None
    dob: Optional[date] = None
    pincode: str
    city: str
    district: str
    state: str
    created_at: datetime
    is_disabled: bool = False
    disability_type: Optional[str] = None
    disability_percentage: Optional[float] = None
    education_level: Optional[str] = None
    screening_status: str = "Pending"
    counseling_status: Optional[str] = None
    counselor_name: Optional[str] = None
    counseling_date: Optional[datetime] = None
    documents_uploaded: List[str] = []
    family_details: Optional[List[dict]] = None
    
    # Assignment fields
    assigned_to_id: Optional[int] = None
    assigned_to_name: Optional[str] = None
    
    # Counseling fields
    feedback: Optional[str] = None
    skills: Optional[List[dict]] = None
    questions: Optional[List[dict]] = None
    workexperience: Optional[List[dict]] = None
    suitable_job_roles: Optional[List[str]] = None
    assigned_to: Optional[List[str]] = None
    remarks: Optional[str] = None
    
    # New Screening fields
    source_of_info: Optional[str] = None
    family_annual_income: Optional[Any] = None
    screening_comments: Optional[str] = None
    screened_by_name: Optional[str] = None
    screening_date: Optional[datetime] = None
    screening_updated_at: Optional[datetime] = None
    screening: Optional[dict] = None
    counseling: Optional[dict] = None
    
    # Flattened Experience & Education
    is_experienced: bool = False
    year_of_experience: Optional[str] = None
    currently_employed: bool = False
    year_of_passing: Optional[int] = None
    
    @model_validator(mode='before')
    @classmethod
    def extract_flattened_data(cls, data: Any) -> Any:
        # data could be ORM object or dict
        is_disabled = False
        disability_type = None
        disability_percentage = None
        education_level = None
        screening_status = "Pending"
        counseling_status = None
        counselor_name = None
        counseling_date = None
        feedback = None
        skills = []
        questions = []
        workexperience = []
        suitable_job_roles = []
        documents_uploaded = []
        family_details = None
        source_of_info = None
        family_annual_income = None
        screened_by_name = None
        screening_date = None
        screening_updated_at = None
        screening_comments = None
        assigned_to_id = None
        assigned_to_name = None
        assigned_to = []
        remarks = None
        
        screening_data = None
        counseling_data = None
        
        # ... (rest of helper functions)
        def get_val(obj, key, default=None):
            if isinstance(obj, dict):
                return obj.get(key, default)
            return getattr(obj, key, default)

        # 1. Disability Data
        details = get_val(data, 'disability_details')
        if details and isinstance(details, dict):
            is_disabled = details.get('is_disabled', False)
            disability_type = details.get('disability_type')
            disability_percentage = details.get('disability_percentage')

        # 1.1 Work Experience Data
        is_experienced = False
        year_of_experience = None
        currently_employed = False
        work_exp = get_val(data, 'work_experience')
        if work_exp and isinstance(work_exp, dict):
            is_experienced = work_exp.get('is_experienced', False)
            year_of_experience = work_exp.get('year_of_experience')
            currently_employed = work_exp.get('currently_employed', False)

        # 2. Education Data
        year_of_passing = None
        edu_details = get_val(data, 'education_details')
        if edu_details and isinstance(edu_details, dict):
            degrees = edu_details.get('degrees', [])
            if degrees and len(degrees) > 0:
                education_level = degrees[0].get('degree_name')
                year_of_passing = degrees[0].get('year_of_passing')

        # 3. Screening Data
        screening = None
        if isinstance(data, dict):
            screening = data.get('screening')
        elif hasattr(data, '__dict__') and 'screening' in data.__dict__:
            screening = data.screening
        
        if screening:
            status_val = get_val(screening, 'status')
            screening_status = status_val if status_val else "In Progress"
            family_details = get_val(screening, 'family_details')
            screening_date = get_val(screening, 'created_at')
            screening_updated_at = get_val(screening, 'updated_at')
            
            # Extract from others JSON field
            others = get_val(screening, 'others')
            if others:
                if isinstance(others, dict):
                    source_of_info = others.get('source_of_info')
                    family_annual_income = others.get('family_annual_income')
                    # Check for 'reason' first (as per user request), then 'comments'
                    screening_comments = others.get('reason')
                
                # Expose full others in screening object for dynamic fields
                screening_data = {'others': others}
            
            # Extract screened_by info
            screened_by = None
            if isinstance(screening, dict):
                screened_by = screening.get('screened_by')
            elif hasattr(screening, '__dict__'):
                # Handle SQLAlchemy relationship carefully
                try:
                    screened_by = screening.screened_by
                except:
                    screened_by = None
            
            if screened_by:
                fname = get_val(screened_by, 'full_name')
                uname = get_val(screened_by, 'username')
                screened_by_name = fname if fname else uname

        # 4. Counseling Data (Relationship)
        counseling = None
        if isinstance(data, dict):
            counseling = data.get('counseling')
        elif hasattr(data, '__dict__') and 'counseling' in data.__dict__:
            counseling = data.counseling

        if counseling:
            counseling_status = get_val(counseling, 'status')
            counseling_date = get_val(counseling, 'counseling_date')
            counselor_name = get_val(counseling, 'counselor_name')
            feedback = get_val(counseling, 'feedback')
            skills = get_val(counseling, 'skills', [])
            questions = get_val(counseling, 'questions', [])
            workexperience = get_val(counseling, 'workexperience', [])
            
            # Extract suitable job roles, assigned_to, and remarks (from property or others)
            suitable_job_roles = get_val(counseling, 'suitable_job_roles', [])
            assigned_to = get_val(counseling, 'assigned_to')
            remarks = get_val(counseling, 'remarks')
            
            if assigned_to and isinstance(assigned_to, str):
                assigned_to = [assigned_to]
            elif not assigned_to:
                assigned_to = []

            c_others = get_val(counseling, 'others')
            if c_others and isinstance(c_others, dict):
                if not suitable_job_roles:
                    suitable_job_roles = c_others.get('suitable_job_roles', [])
                if not assigned_to or len(assigned_to) == 0:
                    val = c_others.get('assigned_to')
                    if isinstance(val, str):
                        assigned_to = [val]
                    else:
                        assigned_to = val or []
                if not remarks:
                    remarks = c_others.get('remarks')
            
            # Extract others for dynamic fields
            counseling_others = get_val(counseling, 'others')
            if counseling_others:
                counseling_data = {'others': counseling_others}
            
            counselor = None
            if isinstance(counseling, dict):
                counselor = counseling.get('counselor')
            elif hasattr(counseling, '__dict__') and 'counselor' in counseling.__dict__:
                counselor = counseling.counselor
            
            if counselor:
                fname = get_val(counselor, 'full_name')
                if fname:
                    counselor_name = fname

        # 5. Documents (Relationship)
        docs_list = None
        if isinstance(data, dict):
            docs_list = data.get('documents')
        elif hasattr(data, '__dict__') and 'documents' in data.__dict__:
            docs_list = data.documents

        if docs_list:
            documents_uploaded = [get_val(d, 'document_type') for d in docs_list]

        # 6. Assignment Data
        assignment = None
        if isinstance(data, dict):
            assignment = data.get('assignment')
        elif hasattr(data, '__dict__') and 'assignment' in data.__dict__:
            assignment = data.assignment
        
        if assignment:
            assigned_to_id = get_val(assignment, 'user_id')
            user = None
            if isinstance(assignment, dict):
                user = assignment.get('user')
            elif hasattr(assignment, '__dict__'):
                try:
                    user = assignment.user
                except:
                    user = None
            
            if user:
                fname = get_val(user, 'full_name')
                uname = get_val(user, 'username')
                assigned_to_name = fname if fname else uname

        # 7. Build response dict
        if isinstance(data, dict):
            data.update({
                'is_disabled': is_disabled,
                'disability_type': disability_type,
                'disability_percentage': disability_percentage,
                'education_level': education_level,
                'screening_status': screening_status,
                'counseling_status': counseling_status,
                'counselor_name': counselor_name,
                'counseling_date': counseling_date,
                'feedback': feedback,
                'skills': skills,
                'questions': questions,
                'workexperience': workexperience,
                'suitable_job_roles': suitable_job_roles,
                'documents_uploaded': documents_uploaded,
                'family_details': family_details,
                'source_of_info': source_of_info,
                'family_annual_income': family_annual_income,
                'screening_comments': screening_comments,
                'screened_by_name': screened_by_name,
                'screening_date': screening_date,
                'screening_updated_at': screening_updated_at,
                'screening': screening_data,
                'counseling': counseling_data,
                'assigned_to': assigned_to,
                'remarks': remarks,
                'is_experienced': is_experienced,
                'year_of_experience': year_of_experience,
                'currently_employed': currently_employed,
                'year_of_passing': year_of_passing,
                'assigned_to_id': assigned_to_id,
                'assigned_to_name': assigned_to_name
            })
            return data
            
        # For ORM objects, create a complete dict
        return {
            'id': get_val(data, 'id'),
            'public_id': get_val(data, 'public_id'),
            'name': get_val(data, 'name'),
            'gender': get_val(data, 'gender'),
            'email': get_val(data, 'email'),
            'phone': get_val(data, 'phone'),
            'whatsapp_number': get_val(data, 'whatsapp_number'),
            'dob': get_val(data, 'dob'),
            'pincode': get_val(data, 'pincode'),
            'city': get_val(data, 'city'),
            'district': get_val(data, 'district'),
            'state': get_val(data, 'state'),
            'created_at': get_val(data, 'created_at'),
            'is_disabled': is_disabled,
            'disability_type': disability_type,
            'disability_percentage': disability_percentage,
            'education_level': education_level,
            'screening_status': screening_status,
            'counseling_status': counseling_status,
            'counselor_name': counselor_name,
            'counseling_date': counseling_date,
            'feedback': feedback,
            'skills': skills,
            'questions': questions,
            'workexperience': workexperience,
            'suitable_job_roles': suitable_job_roles,
            'assigned_to': assigned_to,
            'remarks': remarks,
            'documents_uploaded': documents_uploaded,
            'family_details': family_details,
            'source_of_info': source_of_info,
            'family_annual_income': family_annual_income,
            'screening_comments': screening_comments,
            'screened_by_name': screened_by_name,
            'screening_date': screening_date,
            'screening_updated_at': screening_updated_at,
            'screening': screening_data,
            'counseling': counseling_data,
            'is_experienced': is_experienced,
            'year_of_experience': year_of_experience,
            'currently_employed': currently_employed,
            'year_of_passing': year_of_passing,
            'assigned_to_id': assigned_to_id,
            'assigned_to_name': assigned_to_name
        }



    class Config:
        from_attributes = True


class CandidateStats(BaseModel):
    total: int
    male: int
    female: int
    others: int
    today: int
    weekly: List[int] = []
    screened: int
    not_screened: int
    total_counseled: int = 0
    counseling_pending: int = 0
    counseling_selected: int = 0
    counseling_rejected: int = 0
    docs_total: int = 0
    docs_completed: int = 0
    docs_pending: int = 0
    files_collected: int = 0
    files_to_collect: int = 0
    candidates_fully_submitted: int = 0
    candidates_partially_submitted: int = 0
    candidates_not_submitted: int = 0
    screening_distribution: dict = {}
    counseling_distribution: dict = {}

class ScreeningStats(BaseModel):
    not_screened: int
    screening_distribution: dict = {}


class CandidatePaginatedResponse(BaseModel):
    """Paginated response for candidate listing"""
    items: List[CandidateListResponse]
    total: int

File: frontend/src/models/candidate.ts
// Candidate Screening Types
export interface CandidateScreening {
	id: number;
	candidate_id: number;
	status?: string;
	previous_training?: Record<string, any>;
	documents_upload?: Record<string, any>;
	skills?: Record<string, any>;
	family_details?: any[];
	screened_by_id?: number;
	screened_by?: {
		id: number;
		full_name?: string;
		username: string;
	};
	consent_status?: string;
	others?: {
		willing_for_training?: boolean;
		ready_to_relocate?: boolean;
		source_of_info?: string;
		family_annual_income?: string | number;
		comments?: string;
		[key: string]: any;
	};
	created_at: string;
	updated_at: string;
}

export interface CandidateScreeningCreate {
	status?: string;
	previous_training?: Record<string, any>;
	documents_upload?: Record<string, any>;
	skills?: Record<string, any>;
	family_details?: any[];
	screened_by_id?: number;
	consent_status?: string;
	others?: {
		willing_for_training?: boolean;
		ready_to_relocate?: boolean;
		source_of_info?: string;
		family_annual_income?: string | number;
		comments?: string;
		[key: string]: any;
	};
}

// Candidate Document Types
export interface CandidateDocument {
	id: number;
	candidate_id: number;
	document_type: 'resume' | 'trainer_resume' | 'disability_certificate' | '10th_certificate' | '12th_certificate' | 'degree_certificate' | 'pan_card' | 'aadhar_card' | 'passport_photo' | 'other';
	document_name: string;
	file_path: string;
	file_size?: number;
	mime_type?: string;
	description?: string;
	document_source: 'candidate' | 'trainer';
	is_active: boolean;
	created_at: string;
	updated_at: string;
}

export interface CandidateDocumentCreate {
	document_type: 'resume' | 'trainer_resume' | 'disability_certificate' | '10th_certificate' | '12th_certificate' | 'degree_certificate' | 'pan_card' | 'aadhar_card' | 'passport_photo' | 'other';
	document_name: string;
	file_path: string;
	file_size?: number;
	mime_type?: string;
	description?: string;
	document_source?: 'candidate' | 'trainer';
	is_active?: boolean;
}

// Candidate Counseling Types
export interface CounselingSkill {
	name: string;
	level: 'Beginner' | 'Intermediate' | 'Advanced';
}

export interface CounselingQuestion {
	question: string;
	answer: string;
}

export interface CounselingWorkExperience {
	job_title?: string;
	company?: string;
	years_of_experience?: string;
	currently_working?: boolean;
}

export interface CandidateCounseling {
	id: number;
	candidate_id: number;
	skills?: CounselingSkill[];
	feedback?: string;
	questions?: CounselingQuestion[];
	others?: Record<string, any>;
	workexperience?: CounselingWorkExperience[];
	counselor_name?: string;
	suitable_job_roles?: string[];
	assigned_to?: string[];
	remarks?: string;
	status: 'pending' | 'selected' | 'rejected';
	sub_status?: string;
	counselor_id?: number;
	counseling_date?: string;
	created_at: string;
	updated_at: string;
}

export interface CandidateCounselingCreate {
	skills?: CounselingSkill[];
	feedback?: string;
	questions?: CounselingQuestion[];
	others?: Record<string, any>;
	workexperience?: CounselingWorkExperience[];
	counselor_name?: string;
	suitable_job_roles?: string[];
	assigned_to?: string[];
	remarks?: string;
	status?: 'pending' | 'selected' | 'rejected';
	sub_status?: string;
	counseling_date?: string;
}

export interface CandidateCounselingUpdate extends Partial<CandidateCounselingCreate> { }

export interface Degree {
	degree_name: string;
	specialization: string;
	college_name: string;
	year_of_passing: number;
	percentage: number;
}

export interface EducationDetails {
	degrees: Degree[];
}

export interface DisabilityDetails {
	is_disabled: boolean;
	disability_type?: string;
	disability_percentage?: number;
}

export interface GuardianDetails {
	parent_name?: string;
	relationship?: string;
	parent_phone?: string;
}

export interface WorkExperience {
	is_experienced: boolean;
	currently_employed: boolean;
	year_of_experience?: string;
}

// Main Candidate Types
export interface Candidate {
	id: number;
	public_id: string; // UUID - Used for all API operations
	name: string;
	gender: string;
	dob?: string;
	email: string;
	phone: string;
	whatsapp_number?: string;
	guardian_details?: GuardianDetails;
	pincode: string;
	work_experience?: WorkExperience;
	education_details?: EducationDetails;
	disability_details?: DisabilityDetails;
	city: string;
	district: string;
	state: string;
	created_at: string;
	updated_at: string;
	other?: Record<string, any>;

	// Optional nested relationships (from with_details=true)
	screening?: CandidateScreening;
	documents?: CandidateDocument[];
	counseling?: CandidateCounseling;
}

// Simplified list response (without nested data)
export interface CandidateListItem {
	id: number;
	public_id: string;
	name: string;
	gender: string;
	email: string;
	phone: string;
	whatsapp_number?: string;
	dob?: string;
	pincode: string;
	city: string;
	district: string;
	state: string;
	created_at: string;
	is_disabled?: boolean;
	disability_type?: string;
	education_level?: string;
	specialization?: string;
	screening_status?: string;
	consent_status?: string;
	counseling_status?: string;
	sub_status?: string;
	counselor_name?: string;

	counseling_date?: string;
	feedback?: string;
	skills?: CounselingSkill[];
	questions?: CounselingQuestion[];
	workexperience?: CounselingWorkExperience[];
	documents_uploaded?: string[];
	family_details?: any[];
	source_of_info?: string;
	family_annual_income?: string | number;
	screened_by_name?: string;
	screening_date?: string;
	screening_updated_at?: string;
	disability_percentage?: number;
	screening_comments?: string;
	screening?: CandidateScreening;
	assigned_to?: string[];
	remarks?: string;
	counseling?: CandidateCounseling;

	is_experienced?: boolean;
	year_of_experience?: string;
	currently_employed?: boolean;
	year_of_passing?: number;
	assigned_to_id?: number;
	assigned_to_name?: string;
	registration_type?: string;
}


export interface CandidateCreate extends Omit<Candidate, 'id' | 'public_id' | 'city' | 'district' | 'state' | 'created_at' | 'updated_at' | 'screening' | 'documents' | 'counseling'> {
	city?: string;
	district?: string;
	state?: string;
}

export interface CandidateCheck {
	email: string;
	phone: string;
	pincode: string;
	city?: string;
	district?: string;
	state?: string;
	exclude_public_id?: string;
}

export interface CandidateUpdate {
	name?: string;
	other?: Record<string, any>;
	gender?: string;
	email?: string;
	phone?: string;
	whatsapp_number?: string;
	pincode?: string;
	guardian_details?: GuardianDetails;
	work_experience?: WorkExperience;
	education_details?: EducationDetails;
	disability_details?: DisabilityDetails;
	dob?: string;
	city?: string;
	district?: string;
	state?: string;
}

export interface CandidateStats {
	total: number;
	male: number;
	female: number;
	others: number;
	today: number;
	weekly: number[];
	screened: number;
	not_screened: number;
	total_counseled: number;
	counseling_pending: number;
	counseling_selected: number;
	counseling_rejected: number;
	docs_total: number;
	docs_completed: number;
	docs_pending: number;
	files_collected: number;
	files_to_collect: number;
	files_pending?: number;
	candidates_fully_submitted: number;
	candidates_partially_submitted: number;
	candidates_not_submitted: number;
	pwd_candidates?: number;
	pwd_files_collected?: number;
	pwd_files_to_collect?: number;
	pwd_files_pending?: number;
	non_pwd_candidates?: number;
	non_pwd_files_collected?: number;
	non_pwd_files_to_collect?: number;
	non_pwd_files_pending?: number;
	// Stage 1: All screened candidates — resume + consent_form (Non-PwD=2, PwD=3)
	stage1_total?: number;
	stage1_pwd_count?: number;
	stage1_non_pwd_count?: number;
	stage1_files_collected?: number;
	stage1_files_to_collect?: number;
	stage1_files_pending?: number;
	stage1_pwd_files_collected?: number;
	stage1_pwd_files_to_collect?: number;
	stage1_pwd_files_pending?: number;
	stage1_non_pwd_files_collected?: number;
	stage1_non_pwd_files_to_collect?: number;
	stage1_non_pwd_files_pending?: number;
	stage1_fully_submitted?: number;
	stage1_partially_submitted?: number;
	stage1_not_submitted?: number;
	// Stage 2: Counseling-selected candidates — full 9/10 docs
	stage2_total?: number;
	stage2_pwd_count?: number;
	stage2_non_pwd_count?: number;
	stage2_files_collected?: number;
	stage2_files_to_collect?: number;
	stage2_files_pending?: number;
	stage2_pwd_files_collected?: number;
	stage2_pwd_files_to_collect?: number;
	stage2_pwd_files_pending?: number;
	stage2_non_pwd_files_collected?: number;
	stage2_non_pwd_files_to_collect?: number;
	stage2_non_pwd_files_pending?: number;
	stage2_fully_submitted?: number;
	stage2_partially_submitted?: number;
	stage2_not_submitted?: number;
	screening_distribution?: Record<string, number>;
	counseling_distribution?: Record<string, number>;
	in_training: number;
	moved_to_placement: number;
	got_job: number;
}

export interface ScreeningStats {
	not_screened: number;
	screening_distribution: Record<string, number>;
}

export interface CandidatePaginatedResponse {
	items: CandidateListItem[];
	total: number;
}
